feat(openstack-aws): Add owner/lifetime info in VM's metadata

kaleemsiddiqu opened the pull request:
Owner and lifetime info will be added into VM's metadata for Openstack/AWS providers. This info will give info who owns the VM and what is the lifetime of the VM which will be used for notifying owners and cleaning up the resources based on lifetime of the VMs.
If owner info not fetched, user notified with custom error message for follow up action to take up.

Signed-off-by: ksiddiqu <ksiddiqu@redhat.com>

Tiboris commented:
Just a note also a title of PR must comply with commit message check i have chenged it a bit.
Fist note if possible could you divide this to commit per provider change?

pvoborni commented on src/mrack/transformers/transformer.py:
```diff
@@ -155,3 +162,37 @@ def create_host_requirements(self):
+        mrack_config = global_context.CONFIG
+        require_owner = mrack_config.require_owner()
+        logger.info(f"owner_info : {require_owner}")
+        if require_owner.lower() == "true":
```
required_owner is a boolean (resp. its default value) -> cannot be lowered as string, thus this line fails with:
```
2022-11-04T14:39:56       File "/usr/local/lib/python3.10/site-packages/mrack/transformers/transformer.py", line 172, in validate_ownership_lifetime
2022-11-04T14:39:56         if require_owner.lower() == "true":
2022-11-04T14:39:56     AttributeError: 'bool' object has no attribute 'lower'
```

pvoborni replied:
I'd recommend to modify `require_owner` getter in config.py tp ensure that a bool is returned so that consumers cant rely on it being a boolean.

kaleemsiddiqu replied:
Done, please have look in updated code.

kaleemsiddiqu commented on src/mrack/transformers/transformer.py:
```diff
@@ -15,8 +15,13 @@
+import sys
 import typing
 
+import gssapi
```
IMO should be optional dependency as will be used in specific runtime env. 
So for this we will be modify setup.py with extras_require?

Tiboris replied:
I think for now we should keep it in requirements.txt. Later when we will update the python package to have subpackages we can think of this. I will link above comment to issue I have created so we do not forget.

Tiboris replied:
https://github.com/neoave/mrack/issues/213#issuecomment-1309915790

pvoborni replied:
There is also rpm packaging part. Bigger question is what package should depend on this. The way it is written indicates that python3-mracklib should get a new dependency and thus it would need to be reflected in mrack.spec.  Alternative way is to make it optional and only recommend it in mrack.spec.

kaleemsiddiqu commented:
> First of all thanks for the PR very much. Let us re-visit changes.
> 
> Just a note also a title of PR must comply with commit message check i have changed it a bit. Fist note if possible could you divide this to commit per provider change?

Updated code in two commits

kaleemsiddiqu commented on src/mrack/transformers/transformer.py:
```diff
@@ -155,3 +162,37 @@ def create_host_requirements(self):
         )
         return reqs
+
+    def validate_ownership_lifetime(self, host):
```
Done, please have look in updated code

kaleemsiddiqu commented on src/mrack/transformers/transformer.py:
```diff
@@ -155,3 +162,37 @@ def create_host_requirements(self):
+                        f"{log_msg_start} krb5 cache does not exits, "
+                        f"check and do kinit before provisioning"
+                    )
+                    sys.exit(1)
```
Done, please have look in updated code

kaleemsiddiqu commented on src/mrack/transformers/transformer.py:
```diff
@@ -134,6 +139,8 @@ def validate_config(self):
     def validate_host(self, host):
         """Validate host input that it contains everything needed by provider."""
         log_msg_start = f"{self.dsp_name} [{host.get('name')}]"
+        # ownership and lifetime check
```
Done, please have look in updated code

Tiboris commented on src/mrack/transformers/transformer.py:
```diff
@@ -155,3 +160,65 @@ def create_host_requirements(self):
+        """Check if ownership and lifetime exists for host."""
+        log_msg_start = f"{self.dsp_name} [{host.get('name')}]"
+        mrack_config = global_context.CONFIG
+        require_owner = mrack_config.require_owner()
```
We need to have a look to tests https://github.com/neoave/mrack/pull/207/checks?check_run_id=9356712302 :
```
    def validate_ownership_and_lifetime(self, host):
        """Check if ownership and lifetime exists for host."""
        log_msg_start = f"{self.dsp_name} [{host.get('name')}]"
        mrack_config = global_context.CONFIG
&gt;       require_owner = mrack_config.require_owner()
E       AttributeError: 'NoneType' object has no attribute 'require_owner'
```

kaleemsiddiqu replied:
Fixed

Tiboris replied:
Great thanks @kaleemsiddiqu

pvoborni commented on src/mrack/providers/aws.py:
```diff
@@ -370,6 +370,9 @@ async def create_server(self, req):
             }
 
+        if specs.get("metadata"):
+            request["UserData"] = str(specs.get("metadata"))
```
We should use tags for this feature, not UserData. UserDarta are data provided to e.g. cloud-init so that the vm itself can work with them - not our use case.

pvoborni commented on src/mrack/transformers/transformer.py:
```diff
@@ -155,3 +160,65 @@ def create_host_requirements(self):
         )
         return reqs
+
+    def get_vm_owner(self):
```
An idea for optimization. We might wrap the method with `@cache` decorator to memoize the value and thus run the logic only once. 

https://docs.python.org/3/library/functools.html#functools.cache

kaleemsiddiqu replied:
Added @cache as well.
Also added gssapi in requirements.txt for now as per discussion with @Tiboris 
Based on that marking this conversation resolved.

pvoborni reviewed (changes requested):
Comments added, but in general:
- I'd avoid excessive writes on "info" level. We could be more chatty on debug level, but still on the extend that it is usable for somebody who is investigating an issue.
- methods should not only what their name says
- unit tests needs fixing (I didn't look to deep in it, but probably global context is not initialized)
- you might check with @Tiboris about the gssapi import - it being optional

Tiboris commented:
@kaleemsiddiqu  could you please also add line 
```
require-owner = true
```

to the https://github.com/neoave/mrack/blob/main/src/mrack/data/mrack.conf ?

pvoborni reviewed (commented):
I've tested various scenarios and it works well.

So the only remaining part is the `mrack.spec` discussion. Otherwise it is good from my PoV. 

Doc changes can be done in a separate MR.

Tiboris commented:
Please add
```
Requires:       python3-gssapi
```
after line https://github.com/neoave/mrack/blob/main/mrack.spec#L46

Tiboris reviewed (approved):
Thanks for the patch @kaleemsiddiqu fixes and patience lets ship this ! :)
